src/passes/OptimizeCasts.cpp

```diff
@@ -102,6 +121,268 @@ namespace wasm {
+  // LocalGet, it does not appear here. In the normal case, only one RefCast
+  // needs to be moved to a LocalGet. If a LocalGet is cast to multiple types
+  // which are not subtypes of each other the a trap is inevitable, and we
+  // assume this would already be optimized away beforehand, so we don't care
+  // about this special case.
+  std::unordered_map<LocalGet*, RefCast*> refCastToApply;
+
+  // Maps LocalGets to a RefAs to move to it, to be used by the
+  // EarlyCastApplier. As of right now RefAsNonNull is the only non-extern cast,
+  // so we only have one type of RefAs cast to move.
+  std::unordered_map<LocalGet*, RefAs*> refAsToApply;
+
+  EarlyCastFinder(PassOptions options, Module* module, Function* func)
+    : options(options), numLocals(func->getNumLocals()),
+      currRefCastMove(func->getNumLocals()),
+      currRefAsMove(func->getNumLocals()), testRefCast(options, *module),
+      testRefAs(options, *module) {
+
+    // TODO: generalize this when we handle more than RefAsNonNull
+    RefCast dummyRefCast(module->allocator);
+    RefAs dummyRefAs(module->allocator);
+    dummyRefAs.op = RefAsNonNull;
+
+    testRefCast.visit(&dummyRefCast);
+    testRefAs.visit(&dummyRefAs);
+  }
+
+  void flushRefCastResult(size_t index, Module& module) {
+    if (currRefCastMove[index].target) {
+      if (currRefCastMove[index].bestCast) {
+        // If the fallthrough is equal to the target, this means the cast is
+        // already at the target local.get and doesn't need to be duplicated
+        // again.
+        auto* fallthrough = Properties::getFallthrough(
+          currRefCastMove[index].bestCast, options, module);
+        if (fallthrough != currRefCastMove[index].target) {
+          refCastToApply[currRefCastMove[index].target] =
+            currRefCastMove[index].bestCast;
+        }
+        currRefCastMove[index].bestCast = nullptr;
+      }
+      currRefCastMove[index].target = nullptr;
+    }
+  }
+
+  void flushRefAsResult(size_t index, Module& module) {
+    if (currRefAsMove[index].target) {
+      if (currRefAsMove[index].bestCast) {
+        // As in flushRefCastResult, we need to check if the cast is already at
+        // the target and thus does not need to be moved.
+        auto* fallthrough = Properties::getFallthrough(
+          currRefAsMove[index].bestCast, options, *getModule());
+        if (fallthrough != currRefAsMove[index].target) {
+          refAsToApply[currRefAsMove[index].target] =
+            currRefAsMove[index].bestCast;
+        }
+        currRefAsMove[index].bestCast = nullptr;
+      }
+      currRefAsMove[index].target = nullptr;
+    }
+  }
+
+  static void doNoteNonLinear(EarlyCastFinder* self, Expression**) {
+    for (size_t i = 0; i < self->numLocals; i++) {
+      self->flushRefCastResult(i, *self->getModule());
+      self->flushRefAsResult(i, *self->getModule());
+    }
+  }
+
+  void visitFunction(Function* curr) {
+    for (size_t i = 0; i < numLocals; i++) {
```

Comment:
This can actually just call `doNoteNonLinear` to avoid repeating the code. Or perhaps both can call a helper with a nice name, like `flushAll()`.

Reply:
Done.